File: server/app/schemas/roles.py
"""Role schemas."""

from pydantic import BaseModel

from app.schemas.base import CamelModel


class CriterionResponse(CamelModel):
    id: str
    name: str
    description: str
    weight: int
    question_count: int
    color: str
    sort_order: int


class RoleResponse(CamelModel):
    id: str
    title: str
    department: str
    seniority: str
    location: str
    description: str | None = None
    status: str
    candidate_count: int = 0
    avg_score: float = 0.0
    days_ago: int = 0


class RoleDetailResponse(RoleResponse):
    criteria: list[CriterionResponse] = []


class CriterionInput(BaseModel):
    name: str
    description: str = ""
    weight: int = 20
    question_count: int = 3
    color: str = "#7C6FFF"


class RoleCreateRequest(BaseModel):
    title: str
    department: str
    seniority: str
    location: str
    description: str | None = None
    criteria: list[CriterionInput] = []


class RoleUpdateRequest(BaseModel):
    title: str | None = None
    department: str | None = None
    seniority: str | None = None
    location: str | None = None
    description: str | None = None
    status: str | None = None


class RoleListResponse(CamelModel):
    data: list[RoleResponse]
    count: int

File: client/src/api/roles.ts
import client from "./client";
import type { Role } from "../types";

export interface RoleDetail extends Role {
  criteria: Array<{
    id: string;
    name: string;
    description: string;
    weight: number;
    questionCount: number;
    color: string;
    sortOrder: number;
  }>;
}

export interface RoleCreatePayload {
  title: string;
  department: string;
  seniority: string;
  location: string;
  description?: string;
  criteria: Array<{
    name: string;
    description: string;
    weight: number;
    question_count: number;
    color: string;
  }>;
}

export const rolesApi = {
  list: () =>
    client
      .get<{ data: Role[]; count: number }>("/api/roles")
      .then((r) => r.data.data),
  get: (id: string) =>
    client.get<RoleDetail>(`/api/roles/${id}`).then((r) => r.data),
  create: (payload: RoleCreatePayload) =>
    client.post<RoleDetail>("/api/roles", payload).then((r) => r.data),
  update: (id: string, payload: Partial<Role>) =>
    client.put<Role>(`/api/roles/${id}`, payload).then((r) => r.data),
  delete: (id: string) => client.delete(`/api/roles/${id}`),
};
